---
title: 'fp-ts로 Typescript 함수형 프로그래밍 시작하기 0'
date: 2021-04-20 00:00:04
category: 'TypeScript'
draft: true
---

본 포스트는 fp-ts 공식문서의 [Learning Resources](https://gcanti.github.io/fp-ts/learning-resources/)에 있는 Getting Started에서 소개하는 문서들을 번역하며 학습한 문서입니다. 원본 문서는 [링크](https://dev.to/gcanti/interoperability-with-non-functional-code-using-fp-ts-432e)에서 확인할 수 있으며 작성한 코드들은 [여기](https://github.com/alstn2468/getting-started-fp-ts/tree/main/src/0_interoperability)에서 확인할 수 있습니다.

## fp-ts를 이용한 비함수형 코드와의 상호 운용성

때로는 함수형 프로그래밍 스타일로 작성되지 않은 코드와 상호 운영해야하는 경우가 있습니다. 그러한 상황에서 fp-ts를 이용해 해결하는 방법을 살펴보겠습니다.

## Sentinels

- 유스 케이스: 실패할 수 있고 코드 도메인의 특수한 값을 반환할 수 있는 API
- 예시: `Array.prototype.findIndex`
- 해결 방법: `Option`

```typescript
import { Option, none, some } from 'fp-ts/Option';

function findIndex<A>(
  as: Array<A>,
  predicate: (a: A) => boolean
): Option<number> {
  const index = as.findIndex(predicate);
  return index === -1 ? none : some(index);
}
```

### findIndex 함수 테스트하기

`Array.prototype.findIndex`는 값을 찾지 못했을 경우 `-1`이라는 특수한 값을 반환합니다.

`Array.prototype.findIndex` 함수를 `Option` 타입을 이용해 fp-ts와 함께 사용할 수 있도록 변경했다. `-1`이 반환될 경우 `none`을 반환하고 그렇지 않은 경우 `some`을 반환하도록 작성했습니다.

작성한 `findIndex` 함수를 테스트하기 위한 패키지는 아래와 같습니다.

- `findIndex` 함수를 테스트하기 위한 패키지

```typescript
import * as N from 'fp-ts/number';
import { findIndex } from '../findIndex';
import { isSome, none, some, getEq, Option, isNone } from 'fp-ts/lib/Option';
```

- `findIndex` 함수를 테스트하는 코드

테스트코드에서는 `findIndex`가 `predicate`의 조건에 맞는 값을 찾았을 경우 `isSome` 함수를 이용해 `some`인지 확인한 후 해당 값의 인덱스를 `some`으로 감싸 반환된 값과 같은지 확인합니다.

조건에 맞는 값을 찾지 못했을 경우 `isNone` 함수를 이용해 `none`인지 확인한 후 `none`과 같은지 확인합니다.

```typescript
describe('findIndex함수 테스트 (Sentinels)', () => {
  let array: number[] = [1, 2, 3, 5];
  let result: Option<number>;
  const E = getEq(N.Eq);

  it('findIndex함수가 존재하는 값을 찾았을 경우', () => {
    result = findIndex(array, a => a === 1);
    expect(isSome(result)).toBeTruthy();
    expect(E.equals(result, some(0))).toBeTruthy();
  });

  it('findIndex함수가 값을 찾지 못했을 경우', () => {
    result = findIndex(array, a => a === 4);
    expect(isNone(result)).toBeTruthy();
    expect(E.equals(result, none)).toBeTruthy();
  });
});
```

## `undefined`와 `null`

- 유스 케이스: 실패하고 `undefined`나 `null`을 반환할 수 있는 API
- 예시: `Array.prototype.find`
- 해결 방법: `Option`, `fromNullable`

```typescript
import { Option, fromNullable } from 'fp-ts/Option';

export function find<A>(as: Array<A>, predicate: (a: A) => boolean): Option<A> {
  return fromNullable(as.find(predicate));
}
```

### `find`함수 테스트하기

`Array.prototype`에 있는 `find` 함수는 `null`이나 `undefined`와 같은 Falsy값을 반환할 수 있습니다.

`fromNullable` 함수를 이용해 `find` 함수를 감싸 `Option` 타입을 반환하는 함수로 만들었습니다.

작성한 `find` 함수를 테스트하기 위한 패키지는 아래와 같습니다.

- `find` 함수를 테스트하기 위한 패키지

```typescript
import { find } from './find';
import * as N from 'fp-ts/number';
import { isSome, none, some, getEq, Option, isNone } from 'fp-ts/lib/Option';
```

- `find` 함수를 테스트하는 코드

새로 작성한 `find` 함수는 값을 찾았을 경우 `some` 태그를 갖는 값을 반환합니다. 값을 찾았을 경우에는 `isSome` 메서드가 `true`를 반환하는지 확인하고 값이 `some`으로 감싸진 값과 같은지 확인합니다.

반면에 값을 찾지 못했을 경우 `none` 태그를 값는 값을 반환합니다. 값을 찾지 못했을 경우에는 `isNone` 메서드가 `true`를 반환하는지 확인하고 값이 `none`과 같은지 확인합니다.

```typescript
describe('find함수 테스트 (undefined와 null)', () => {
  let array: number[] = [1, 2, 3, 5];
  let result: Option<number>;
  const E = getEq(N.Eq);

  it('find함수가 존재하는 값을 찾았을 경우', () => {
    result = find(array, a => a === 1);
    expect(isSome(result)).toBeTruthy();
    expect(E.equals(result, some(1))).toBeTruthy();
  });

  it('find함수가 값을 찾지 못했을 경우', () => {
    result = find(array, a => a === 4);
    expect(isNone(result)).toBeTruthy();
    expect(E.equals(result, none)).toBeTruthy();
  });
});
```

## 예외

- 유즈 케이스: `throw`될 수 있는 API
- 예시: `JSON.parse`
- 해결 방법: `Eiter`, `tryCatch`

```typescript
import { Either, tryCatch } from 'fp-ts/Either';

function parse(s: string): Either<Error, unknown> {
  return tryCatch(
    () => JSON.parse(s),
    reason => new Error(String(reason))
  );
}
```

### `parse` 함수 테스트하기

`JSON.parse` 함수는 인자로 잘못된 문자열이 전달될 경우 예외를 발생시킵니다. fp-ts의 `tryCatch`를 이용하면 fp-ts와 함께 사용할 수 있도록 `Either` 타입을 반환시킬 수 있습니다.

작성한 `parse` 함수를 테스트하기 위한 패키지는 아래와 같습니다.

- `parse` 함수를 테스트하기 위한 패키지

```typescript
import { parse } from '../parse';
import { isLeft, isRight, getOrElse } from 'fp-ts/Either';
```

- `parse` 함수를 테스트하는 코드

`parse` 함수가 성공하는 값과 실패하는 값을 준비 한 후 `parse` 함수가 정상적으로 실헹되는 경우 `Either` 패키지에 있는 `isRight` 함수를 사용해 정상적으로 코드가 실행되었는지 확인합니다. 그 후 `getOrElse` 함수를 이용해 `right`에서 값을 꺼내 기대한 객체와 값이 같은지 확인합니다.

반면에 실패했을 경우에는 `isLeft` 함수를 이용해 `left`가 반환되었는지 확인합니다. `getOrElse` 함수의 인자로 전달된 함수가 반환한 값과 기대하는 값이 같은지 확인합니다.

```typescript
describe('parse함수 테스트 (예외)', () => {
  const success = '{"a": 1, "b": 2}';
  const fail = '{"a": 1, "b"}';
  let result;
  it('parse함수가 정상적으로 실행됐을 경우', () => {
    result = parse(success);
    expect(isRight(result)).toBeTruthy();
    expect(getOrElse(() => ({ error: true }))(result)).toMatchObject({
      a: 1,
      b: 2,
    });
  });

  it('parse함수 실행 중 예외가 발생했을 경우', () => {
    result = parse(fail);
    expect(isLeft(result)).toBeTruthy();
    expect(getOrElse(() => ({ a: 1 }))(result)).toMatchObject({
      a: 1,
    });
  });
});
```

## 무작위 값

- 유즈 케이스: 비결정론적인 값을 반환하는 API
- 예시: `Math.random`
- 해결 방법: `IO`

```typescript
import { IO } from 'fp-ts/IO';

const random: IO<number> = () => Math.random();
```

### `random` 함수 테스트하기

`Math.random` 함수는 절대 실패하지 않지만 비결정론적인 값을 반환합니다. 따라서 fp-ts와 함께 사용하기 위해서 `IO` 타입으로 감싸 값을 반환하면 됩니다.

작성한 `random` 함수를 테스트하기 위한 패키지는 아래와 같습니다.

- `random` 함수를 테스트하기 위한 패키지

```typescript
import { random } from '../random';
```

- `random` 함수를 테스트하는 코드

`Math.random`은 앞에서 설명했듯이 비결정론적인 값을 반환합니다. 따라서 값을 mocking하도록 도와주는 `jest.fn` 함수가 반환한 `jest.Mock` 인스턴스의 `mockReturnValue` 함수를 이용해 고정적으로 값을 반환할 수 있게 해줍니다. 그 후 mocking된 값이 작성한 `random` 함수가 반환하는 값과 같은지 확인합니다.

```typescript
describe('random함수 테스트 (랜덤)', () => {
  let result;
  it('Math.random함수를 mocking해 테스트하기', () => {
    Math.random = jest.fn().mockReturnValue(0.5);
    result = random();
    expect(result).toBe(0.5);
  });
});
```

## 동기 부수 효과

- 유즈 케이스: 전역 상태를 읽거나 쓰는 API
- 예시: `localStorage.getItem`
- 해결 방법: `IO`

```typescript
import { Option, fromNullable } from 'fp-ts/Option';
import { IO } from 'fp-ts/IO';

function getItem(key: string): IO<Option<string>> {
  return () => fromNullable(localStorage.getItem(key));
}
```

### `getItem` 함수 테스트하기

`localStorage.getItem` 함수는 값을 찾지 못할 경우 `null`을 반환하는 함수입니다. 따라서 `fromNullable`을 이용해 값을 찾지 못할 경우 `none`을 반환하고 값을 찾았을 경우 `some`을 반환하도록 구현할 수 있습니다.

작성한 `getItem` 함수를 테스트하기 위한 패키지는 아래와 같습니다.

- `getItem` 함수를 테스트하기 위한 패키지

```typescript
import * as S from 'fp-ts/string';
import { getItem } from '../getItem';
import { isSome, isNone, none, some, getEq, Option } from 'fp-ts/Option';
```

- `getItem` 함수를 테스트하는 코드

`localStorage.getItem` 함수는 브라우저 상에서 사용할 수 있는 함수이므로 `jest.fn`을 이용해 원하는 `key`가 들어왔을 경우에만 값을 반환하도록 mocking해 줍니다. 여기서 주의해야할 점은 구현한 `getItem` 함수는 함수를 반환하는 함수이므로 함수 호출 연산자를 한번 더 사용해야 합니다.

`getItem` 함수가 반환한 함수가 성공적으로 값을 가져왔을 경우 `some`인지 확인한 후 값이 같은지 비교해주고 값을 찾지 못해 `none`을 반환할 경우 `none`인지 확인합니다.

```typescript
describe('getItem함수 테스트 (동기 부수 효과)', () => {
  window.localStorage.__proto__.getItem = jest.fn(key => {
    if (key === 'success') return 'success';
    return null;
  });
  let result: Option<string>;
  const E = getEq(S.Eq);
  it('getItem함수가 값을 정상적으로 가져온 경우', () => {
    result = getItem('success')();
    expect(isSome(result)).toBeTruthy();
    expect(E.equals(result, some('success'))).toBeTruthy();
  });

  it('getItem함수가 값을 가져오지 못했을 경우', () => {
    result = getItem('fail')();
    expect(isNone(result)).toBeTruthy();
    expect(E.equals(result, none)).toBeTruthy();
  });
});
```

- 유즈 케이스: 전역 상태를 읽고 / 쓰고 `throw`할 수있는 API
- 예시: `readFileSync`
- 해결 방법: `IOEither`, `tryCatch`

```typescript
import * as fs from 'fs';
import { IOEither, tryCatch } from 'fp-ts/IOEither';

function readFileSync(path: string): IOEither<Error, string> {
  return tryCatch(
    () => fs.readFileSync(path, 'utf8'),
    reason => new Error(String(reason))
  );
}
```

## 비동기 부수 효과

- 유즈 케이스: 비동기 계산을 수행하는 API
- 예시: 표준 입력으로부터 읽을 때
- 해결 방법: `Task`

```typescript
import { createInterface } from 'readline';
import { Task } from 'fp-ts/Task';

const read: Task<string> = () =>
  new Promise<string>(resolve => {
    const rl = createInterface({
      input: process.stdin,
      output: process.stdout,
    });
    rl.question('', answer => {
      rl.close();
      resolve(answer);
    });
  });
```

- 유즈 케이스: 비동기 계산을 수행하고 거부될 수 있는 API
- 예시: 표준 `fetch`
- 해결 방법: `TaskEither`, `tryCatch`

```typescript
import { TaskEither, tryCatch } from 'fp-ts/TaskEither';

function get(url: string): TaskEither<Error, string> {
  return tryCatch(
    () => fetch(url).then(res => res.text()),
    reason => new Error(String(reason))
  );
}
```
